webapp/src/ts/services/format-data-record.service.ts

```diff
@@ -0,0 +1,716 @@
+    return this.dbService
+      .get()
+      .query('medic-client/contacts_by_reference', options)
+      .then((result) => {
+        if (!result.rows.length) {
+          return;
+        }
+        if (result.rows.length > 1) {
+          console.warn('More than one patient person document for shortcode "' + patientId + '"');
+        }
+        return this.lineage.fetchHydratedDoc(result.rows[0].id);
+      });
+  }
+
+  private fieldsToHtml(
+    settings,
+    doc,
+    keys,
+    labels?,
+    locale?,
+    data?,
+    def?
+  ) {
+    if (!def && doc && doc.form) {
+      def = this.getForm(settings, doc.form);
+    }
+
+    if (_.isString(def)) {
+      def = this.getForm(settings, def);
+    }
+
+    if (!data) {
+      data = Object.assign({}, doc, doc.fields);
+    }
+
+    const fields = {
+      headers: [],
+      data: [],
+    };
+
+    _.forEach(keys, (key) => {
+      if (_.isArray(key)) {
+        const result:any = this.fieldsToHtml(settings, doc, key[1], labels, locale, data[key[0]], def);
+        result.isArray = true;
+        fields.data.push(result);
+        fields.headers.push({ head: this.titleize(key[0]) });
+      } else {
+        const label = labels.shift();
+        fields.headers.push({ head: this.getMessage(settings, label) });
+        if (def && def[key]) {
+          def = def[key];
+        }
+        fields.data.push({
+          isArray: false,
+          value: this.prettyVal(settings, data, key, def, locale),
+          label: label,
+          target: this.getClickTarget(key, doc)
+        });
+      }
+    });
+
+    return fields;
+  }
+
+  private getClickTarget(key, doc) {
+    if (this.patientFields.includes(key)) {
+      const id = (doc.patient && doc.patient._id) ||
+        (doc.fields && doc.fields.patient_uuid);
+      if (id) {
+        return { url: ['/contacts', id] };
+      }
+    } else if (key === 'case_id') {
+      const id = doc.case_id || doc.fields.case_id;
+      if (id) {
+        return { filter: `case_id:${id}` };
+      }
+    }
+  }
+
+  /*
+    * Get an array of keys from the form.  If dot notation is used it will be an
+    * array of arrays.
+  *
+  * @param Object def - form definition
+  *
+  * @return Array  - form field keys based on forms definition
+  */
+  private getFormKeys(def) {
+    const keys = {};
+
+    const getKeys = (key, hash) => {
+      if (key.length > 1) {
+        const tmp = key.shift();
+        if (!hash[tmp]) {
+          hash[tmp] = {};
+        }
+        getKeys(key, hash[tmp]);
+      } else {
+        hash[key[0]] = '';
+      }
+    };
+
+    const hashToArray = (hash) => {
+      const array = [];
+
+      _.forEach(hash, (value, key) => {
+        if (typeof value === 'string') {
+          array.push(key);
+        } else {
+          array.push([key, hashToArray(hash[key])]);
+        }
+      });
+
+      return array;
+    };
+
+    if (def) {
+      Object.keys(def.fields).forEach((key) => {
```

Comment:
I think it's better to put each function in a diff line.

Reply:
I think this code was just copied, but I will :)